### PR TITLE
Adding caching on the Bitbucket Server, fixing NullPointerException 

### PR DESCRIPTION
**Pull request by 2improveIT**

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/CachedLRUObjects.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package com.cloudbees.jenkins.plugins.bitbucket.server.client.cache;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.logging.Logger;
+
+public class CachedLRUObjects<K,V> {
```
Comment (darxriggs) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/CachedLRUObjects.java]: Why implementing a custom cache solution and not using existing ones like Guava's (https://github.com/google/guava/wiki/CachesExplained)?

Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/CachedLRUObjects.java]: https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/91#issuecomment-352396170

As @stephenc said, say no to Guava :sweat:

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/PullRequestSCMHead.java [hunk 2]
```diff
@@ -186,7 +186,11 @@ public ChangeRequestCheckoutStrategy getCheckoutStrategy() {
     @NonNull
     @Override
     public String getOriginName() {
-        return branchName;
+        if (strategy == ChangeRequestCheckoutStrategy.MERGE) {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/PullRequestSCMHead.java]: No, create a new method. :cry:

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java [hunk 3]
```diff
@@ -187,6 +189,17 @@ protected BitbucketSCMSourceRequest(@NonNull final BitbucketSCMSource source,
...
+    }
+
+    /**
+     * Returns {@code true} if notifications are disabled.
+     *
+     * @return {@code true} if notifications are disabled.
+     */
+
+    public final boolean isDisableNotifications() {
+        return disableNotifications;
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java]: Gratitious change, please create another PR

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/BitbucketCache.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+import com.cloudbees.jenkins.plugins.bitbucket.api.*;
+import com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+
+public class BitbucketCache {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/BitbucketCache.java]: Why is this being recreated for the server? When Cloud already one, with some minor changes you could reuse.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/CachedBitbucketApi.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+import com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.concurrent.Callable;
+import jenkins.scm.api.SCMFile;
+
+public class CachedBitbucketApi implements BitbucketApi {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/CachedBitbucketApi.java]: Seems overkill to create another class.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java (hunk 3 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java]: This is actually already supported in a trait plugin: https://plugins.jenkins.io/skip-notifications-trait

Comment (amuniz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java]: A list of available traits (for this plugin) would be helpful. Maybe just in the README.md?

**Comment by stephenc** [comment]
This PR is doing too many things. I would prefer if it was split out. I shall defer to @casz and @amuniz on whether splitting is required

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java [hunk 6]
```diff
@@ -187,6 +193,17 @@ protected BitbucketSCMSourceRequest(@NonNull final BitbucketSCMSource source,
         repoOwner = source.getRepoOwner();
         repository = source.getRepository();
         pullRequestData = new HashMap<>();
+        disableNotifications = context.notificationsDisabled();
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSourceRequest.java]: Please remove there is already a trait plugin that provides this support: https://plugins.jenkins.io/skip-notifications-trait

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/PullRequestSCMHead.java (hunk 2 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/PullRequestSCMHead.java]: This does not work for cloud: https://stackoverflow.com/questions/35216431/bitbucket-pull-request-refspec

This line will be broken
https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/169/files#diff-f47f4b30aafc3302f647c3dbcec3ac48L159

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/BitbucketCache.java (hunk 4 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/cache/BitbucketCache.java]: This already exists: https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/be28d7167c4213fc57bb8fd6982484367dcb5885/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/Cache.java

**Pull request closed without merging**